src/Unifier.ml

```diff
@@ -218,12 +301,23 @@ and unify_descriptors t desc1 desc2 =
      raise UnifySkolemInternal
 
   | _, _ ->
-     { (* We pick the skolem identifier if there is such *)
-      id        = if desc2.skolem then desc2.id else desc1.id;
-      structure = unify_structures t desc1.structure desc2.structure;
-      rank      = min desc1.rank desc2.rank;
-      skolem    = desc1.skolem || desc2.skolem (* skolemize *)
-     }
+      (* skolemize *)
+     let skolem = desc1.skolem || desc2.skolem in
+     let new_desc =
+       { (* We pick the skolem identifier if there is such *)
+         id          = if desc2.skolem then desc2.id else desc1.id;
+         structure   = unify_structures t desc1.structure desc2.structure;
+         rank        = min desc1.rank desc2.rank;
+         skolem;
+         monomorphic = not skolem && (desc1.monomorphic || desc2.monomorphic)
+       }
+       in
+       if new_desc.monomorphic then
+         (* Propagate the monomorphism to all type vars used on the
+          structure, if it exists. *)
+         monomorphize_structure (VarMap.create 128) new_desc.structure;
```

Comment:
I did, but I had to use tuples for this because of some limitations of what OCaml allows as the lhs of an let rec. In my opinion the result is worse than the code before...

Reply:
Ok, if you say so then I trust this is better.